rosbag2_cpp/src/rosbag2_cpp/player_clock.cpp

```diff
@@ -0,0 +1,99 @@
+// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+#include <mutex>
+#include <thread>
+
+#include "rosbag2_cpp/player_clock.hpp"
+#include "rosbag2_cpp/types.hpp"
+
+namespace
+{
+
+/**
+ * Stores an exact time match between a system steady clock and the semantic play clock.
+ * This is created whenever a factor changes such that a new base reference is needed
+ * such as pause, resume, rate change, or jump
+ */
+struct TimeSync
+{
+  rosbag2_cpp::PlayerClock::PlayerTimePoint player_time;
+  rosbag2_cpp::PlayerClock::RealTimePoint real_time;
```

Comment:
After reviewing logic in this PR I come to conclusion that it will be much clear and easier to understand if these two variables would be named as:
```cpp
  rosbag2_cpp::PlayerClock::RecorderTimePoint recording_start_time;
  rosbag2_cpp::PlayerClock::RealTimePoint playback_start_time;
```
And also `PlayerTimePoint` would be named as `RecorderTimePoint` since it is time point in time when message was really recorder.

Reply:
The "realtime" point is totally unrelated to the player or the recorder, this name change would be incorrect. "real" in this implementation refers to arbitrary system steady clock time, which is never revealed to users, it is only used internally to calculate offsets. 

These will not always be a "start" time. The plan is to create a new reference pair when a change occurs, such as set_rate or resume from pause. This allows us to avoid complexity of accumulating these offsets over time.

I disagree with Recorder naming - this is the "current time according to the player" which may be totally unrelated to the recording, if for example it's using simulated time from an external /clock publisher

Reply:
I renamed `RealTimePoint` to `SteadyTimePoint` and added some documenting comments - so that at least makes this more clear.

Reply:
@emersonknapp How about renaming `player_time` to the `recording_start_time`? \
 I've tried to think about it more in perspective for future development and it seems this variable always will keep `recording_start_time`. Even if we will advance time we will advance `real_time` and keep `player_time` with original "`recording_start_time`". \
I might missed some case which you might have in mind. Please correct me if you aware about such case or we can postpone this question till we will have some progress in feature development and it will be clear if it's going to changed or not. \
Or it could be alternative like `recording_base_time` and `playback_base_time`.

Reply:
This is the way I'm thinking about it - based on the following reasoning I believe we need two movable reference points. One reference point in ROS time (Player time / message time) and one reference point in Steady time.

In the following diagram you can see the flow of Steady Time (`S`), which always continues at the same rate, and the flow of ROS Time (`R`) which may change speed or pause, as two 1 dimensional timelines. The bag we are playing starts at `R0` and goes until `R3`. We start playback rate at 0.5, and we change the rate twice over the course of the playback.

![time_ref](https://user-images.githubusercontent.com/537409/112710870-877d6700-8e81-11eb-8f7b-4af3a1a33939.png)

* At time `S1` we change the rate to `2.0`. At time `S2` we change the rate to `1.0`
* When we want to know `R PlayerClock::now()` - we know `S s_now = now_fn()`
* When we want to `void PlayerClock::sleep_until(R r_until)` 

To figure out the conversions between arbitrary points Rx<->Sx, Ry<->Sy, Rz<->Sz, I see the equations looking like the following

<img width="274" alt="Screen Shot 2021-03-26 at 10 29 39 PM" src="https://user-images.githubusercontent.com/537409/112711089-f4ddc780-8e82-11eb-8b24-65f05a08b8f5.png">

Based on the above, it shows that in order to convert between the two time points we need both of the latest `Si` and `Ri` reference points - I'll call these `S_ref` and `R_ref`

<img width="288" alt="Screen Shot 2021-03-26 at 10 29 58 PM" src="https://user-images.githubusercontent.com/537409/112711111-22c30c00-8e83-11eb-8392-744630200a43.png">

Whenever we change the rate (or pause, or jump) - we need to take a snapshot of the current `S_ref` and `R_ref` values to use in order to calculate the time conversion. The combination of these two values together contains the integration of all past rates, pauses, etc. I don't see how you could do this calculation accurately with only 1 datapoint, I think you need both. Please let me know if I'm missing something that lets us reduce these 2 reference numbers into a single reference number and a `const ros_starting_time`

Reply:
I've opened a PR adding a section to the design, with this information - https://github.com/ros2/rosbag2/pull/700

If you have a counterpoint to this logic, let's discuss it there. Until then, I plan to move forward with this PR so that we can start building the next features on top of it - if we revise the design it should be simple enough to modify the internals of the `PlayerClock` to match the updated design.

Reply:
@emersonknapp Thanks for the explanation. I will try to comprehend your approach and comment in #700 directly. \
And for sure feel free to move forward with this PR.